File: 0-inverted-index-and-bm25-from-scratch/0-typescript/src/search/search.controller.ts
import {
  BadRequestException,
  Body,
  Controller,
  Delete,
  Get,
  Param,
  Post,
  Query,
} from '@nestjs/common';
import { IndexDocumentDto } from './dto';
import { SearchService } from './search.service';

/** REST surface for the BM25 inverted-index service. */
@Controller('api/search')
export class SearchController {
  constructor(private readonly search: SearchService) {}

  // POST /api/search/index -> 201 { id, indexed: true }
  @Post('index')
  async index(@Body() dto: IndexDocumentDto): Promise<{ id: string; indexed: boolean }> {
    if (!dto?.id || !dto?.content) {
      throw new BadRequestException('id and content are required non-empty strings');
    }
    this.search.indexDocument(dto.id, dto.content);
    await this.search.saveSnapshot();
    return { id: dto.id, indexed: true };
  }

  // GET /api/search?q=...&limit=n -> 200 { query, hits }
  @Get()
  query(
    @Query('q') q: string,
    @Query('limit') limit?: string,
  ): { query: string; hits: ReturnType<SearchService['search']> } {
    const parsedLimit = Number(limit) > 0 ? Number(limit) : 5;
    const hits = this.search.search(q ?? '', parsedLimit);
    return { query: q ?? '', hits };
  }

  // GET /api/search/stats -> 200 { numDocs, vocabSize, avgdl }
  @Get('stats')
  stats(): { numDocs: number; vocabSize: number; avgdl: number } {
    return this.search.stats();
  }

  // DELETE /api/search/index/:id -> 200 { id, deleted }
  @Delete('index/:id')
  async remove(@Param('id') id: string): Promise<{ id: string; deleted: boolean }> {
    const deleted = this.search.removeDocument(id);
    await this.search.saveSnapshot();
    return { id, deleted };
  }

  // POST /api/search/reset -> 201 { reset: true }
  @Post('reset')
  async reset(): Promise<{ reset: boolean }> {
    this.search.reset();
    await this.search.saveSnapshot();
    return { reset: true };
  }
}

File: 0-inverted-index-and-bm25-from-scratch/1-java/src/main/java/com/starci/bm25/search/SearchController.java
package com.starci.bm25.search;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.LinkedHashMap;
import java.util.Map;

/** REST surface for the BM25 inverted-index service. */
@RestController
@RequestMapping("/api/search")
public class SearchController {

    private final SearchService service;

    public SearchController(SearchService service) {
        this.service = service;
    }

    /** Request body for POST /api/search/index. */
    public record IndexRequest(String id, String content) {}

    // POST /api/search/index -> 201 { id, indexed: true }
    @PostMapping("/index")
    public ResponseEntity<Map<String, Object>> index(@RequestBody IndexRequest req) {
        if (req.id() == null || req.id().isBlank() || req.content() == null || req.content().isBlank()) {
            Map<String, Object> err = new LinkedHashMap<>();
            err.put("error", "id and content are required non-empty strings");
            return ResponseEntity.badRequest().body(err);
        }
        service.indexDocument(req.id(), req.content());
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("id", req.id());
        body.put("indexed", true);
        return ResponseEntity.status(HttpStatus.CREATED).body(body);
    }

    // GET /api/search?q=...&limit=n -> 200 { query, hits }
    @GetMapping
    public Map<String, Object> search(@RequestParam(name = "q", defaultValue = "") String q,
                                      @RequestParam(name = "limit", defaultValue = "5") int limit) {
        int effective = limit > 0 ? limit : 5;
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("query", q);
        body.put("hits", service.search(q, effective));
        return body;
    }

    // GET /api/search/stats -> 200 { numDocs, vocabSize, avgdl }
    @GetMapping("/stats")
    public Map<String, Object> stats() {
        return service.stats();
    }

    // DELETE /api/search/index/{id} -> 200 { id, deleted }
    @DeleteMapping("/index/{id}")
    public Map<String, Object> remove(@PathVariable String id) {
        boolean deleted = service.removeDocument(id);
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("id", id);
        body.put("deleted", deleted);
        return body;
    }

    // POST /api/search/reset -> 201 { reset: true }
    @PostMapping("/reset")
    public ResponseEntity<Map<String, Object>> reset() {
        service.reset();
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("reset", true);
        return ResponseEntity.status(HttpStatus.CREATED).body(body);
    }
}
